src/dos/dos_locale.cpp

```diff
@@ -2651,723 +384,445 @@ static void refresh_currency_format(const LocaleInfoEntry &source)
+		return std::string(MSG_Get("SCRIPT_LATIN")) + " (Dvorak)";
+	case KeyboardScript::LatinNonStandard:
+		return std::string(MSG_Get("SCRIPT_LATIN")) + " (" +
+		       MSG_Get("SCRIPT_PROPERTY_NON_STANDARD") + ")";
+	case KeyboardScript::Arabic: return MSG_Get("SCRIPT_ARABIC");
+	case KeyboardScript::Armenian: return MSG_Get("SCRIPT_ARMENIAN");
+	case KeyboardScript::Cherokee: return MSG_Get("SCRIPT_CHEROKEE");
+	case KeyboardScript::Cyrillic: return MSG_Get("SCRIPT_CYRILLIC");
+	case KeyboardScript::CyrillicPhonetic:
+		return std::string(MSG_Get("SCRIPT_CYRILLIC")) + " (" +
+		       MSG_Get("SCRIPT_PROPERTY_PHONETIC") + ")";
+	case KeyboardScript::Georgian: return MSG_Get("SCRIPT_GEORGIAN");
+	case KeyboardScript::Greek: return MSG_Get("SCRIPT_GREEK");
+	case KeyboardScript::Hebrew: return MSG_Get("SCRIPT_HEBREW");
+	case KeyboardScript::Thai: return MSG_Get("SCRIPT_THAI");
+	default: assert(false);
+	}
 
-	case Country::Niger: return assert_code_page(30028);
+	return "<unknown keyboard script>";
+}
 
-	case Country::Kazakhstan:
-	case Country::Kyrgyzstan:
-	case Country::Mongolia: return assert_code_page(58152);
+std::string DOS_GetShortcutKeyboardScript1()
+{
+	return MSG_Get("KEYBOARD_SHORTCUT_SCRIPT_1");
+}
 
-	case Country::Azerbaijan: return assert_code_page(58210);
+std::string DOS_GetShortcutKeyboardScript2()
+{
+	return MSG_Get("KEYBOARD_SHORTCUT_SCRIPT_2");
+}
 
-	case Country::Georgia: return assert_code_page(59829);
+std::string DOS_GetShortcutKeyboardScript3()
+{
+	return MSG_Get("KEYBOARD_SHORTCUT_SCRIPT_3");
+}
 
-	case Country::Uzbekistan: return assert_code_page(62306);
+std::optional<KeyboardScript> DOS_GetKeyboardLayoutScript1(const std::string& layout)
+{
+	const auto layout_deduplicated = deduplicate_layout(layout);
 
-	default:
-		LOG_WARNING("DOS: No default code page for country %d, '%s'",
-		            enum_val(country_id),
-		            get_country_name(country_id).c_str());
-		return assert_code_page(DefaultCodePage437);
+	for (const auto& entry : LocaleData::KeyboardLayoutInfo) {
+		assert(!entry.layout_codes.empty());
+		if (entry.layout_codes[0] == layout_deduplicated) {
+			return entry.primary_script;
+		}
 	}
+
+	return {};
 }
 
-std::string DOS_GetLayoutFromHost()
+std::optional<KeyboardScript> DOS_GetKeyboardLayoutScript2(const std::string& layout,
+                                                           const uint16_t code_page)
 {
-#if defined(WIN32)
-	// Use Windows-specific calls to extract the keyboard layout
-
-	WORD current_kb_layout = LOWORD(GetKeyboardLayout(0));
-	WORD current_kb_sub_id = 0;
-	char layout_id_string[KL_NAMELENGTH];
+	const auto layout_deduplicated = deduplicate_layout(layout);
+
+	for (const auto& entry : LocaleData::KeyboardLayoutInfo) {
+		assert(!entry.layout_codes.empty());
+		if (entry.layout_codes[0] == layout_deduplicated) {
+			const auto& list = entry.secondary_scripts;
+			if (list.contains(code_page)) {
+				return list.at(code_page);
+			}
+		}
+	}
```

Comment:
Can these three functions (get keyboard layout script 1, 2, 3) be de-duped using a couple static helpers? Something like:

```c++
// Helper to find keyboard layout entry
static const auto* find_layout_entry(const std::string& layout) {
    const auto layout_deduplicated = deduplicate_layout(layout);
    for (const auto& entry : LocaleData::KeyboardLayoutInfo) {
        assert(!entry.layout_codes.empty());
        if (entry.layout_codes[0] == layout_deduplicated) {
            return &entry;
        }
    }
    return nullptr;
}

// Helper to get script from codepage map
static std::optional<KeyboardScript> get_script_from_map(const std::map<uint16_t, KeyboardScript>& script_map, 
                                                const uint16_t code_page) {
    return script_map.contains(code_page) ? std::optional{script_map.at(code_page)} : std::nullopt;
}
```

Then the three API calls would be:

```c++
std::optional<KeyboardScript> DOS_GetKeyboardLayoutScript1(const std::string& layout) {
    const auto* entry = find_layout_entry(layout);
    return entry ? std::optional{entry->primary_script} : std::nullopt;
}

std::optional<KeyboardScript> DOS_GetKeyboardLayoutScript2(const std::string& layout,
                                                         const uint16_t code_page) {
    const auto* entry = find_layout_entry(layout);
    return entry ? get_script_from_map(entry->secondary_scripts, code_page) : std::nullopt;
}

std::optional<KeyboardScript> DOS_GetKeyboardLayoutScript3(const std::string& layout,
                                                         const uint16_t code_page) {
    const auto* entry = find_layout_entry(layout);
    return entry ? get_script_from_map(entry->tertiary_scripts, code_page) : std::nullopt;
}
```

Reply:
I think introducing the `get_script_from_map` is a little overkill - but I have reworked the code and extracted most of the processing to a separate function.